x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/GetDataFrameAnalyticsStatsAction.java

```diff
@@ -175,11 +182,44 @@ public Stats(StreamInput in) throws IOException {
                 id = in.readString();
                 state = DataFrameAnalyticsState.fromStream(in);
                 failureReason = in.readOptionalString();
-                progressPercentage = in.readOptionalInt();
+                if (in.getVersion().before(Version.V_7_4_0)) {
+                    progress = readProgressFromLegacy(state, in);
+                } else {
+                    progress = in.readList(PhaseProgress::new);
+                }
                 node = in.readOptionalWriteable(DiscoveryNode::new);
                 assignmentExplanation = in.readOptionalString();
             }
 
+            private static List<PhaseProgress> readProgressFromLegacy(DataFrameAnalyticsState state, StreamInput in) throws IOException {
+                Integer legacyProgressPercent = in.readOptionalInt();
+                if (legacyProgressPercent == null) {
+                    return Collections.emptyList();
+                }
+
+                int reindexingProgress = 0;
+                int analyzingProgress = 0;
+                switch (state) {
+                    case ANALYZING:
+                        analyzingProgress = legacyProgressPercent;
+                        break;
+                    case REINDEXING:
+                        reindexingProgress = legacyProgressPercent;
+                        break;
+                    case STARTED:
+                    case STOPPED:
+                    case STOPPING:
+                    default:
+                        return null;
+                }
+
+                return Arrays.asList(
+                    new PhaseProgress("reindexing", reindexingProgress),
+                    new PhaseProgress("loading_data", 100),
```

Comment:
#nit, it seems to me that `loading_data` would not be `100` unless it was currently in the `ANALYZING` state. Though, this does not really matter too much.

Reply:
Good catch. I think I left what I was doing there unfinished.